java/src/org/openqa/selenium/remote/BiDiDelegator.java

```diff
@@ -0,0 +1,82 @@
+// Licensed to the Software Freedom Conservancy (SFC) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The SFC licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.openqa.selenium.remote;
+
+import static java.util.Collections.emptyMap;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import org.openqa.selenium.internal.Require;
+import org.openqa.selenium.remote.http.ClientConfig;
+import org.openqa.selenium.remote.http.HttpClient;
+
+// Delegates the command to the respective command executor.
+// If the commands is supported in BiDi, it is delegated to BiDiCommandExecutor
+// Otherwise, it falls back on HttpCommandExecutor
+// The reason both are required is that, HttpCommandExecutor route will always be used to start and
+// stop the session.
+public class BiDiDelegator extends HttpCommandExecutor implements CommandExecutor {
+
+  private BiDiCommandExecutor biDiCommandExecutor = null;
+  private static final Set<String> biDiCommands = new HashSet<>();
+
+  static {
+    biDiCommands.add(DriverCommand.GET);
+    biDiCommands.add(DriverCommand.PRINT_PAGE);
+  }
+
+  public BiDiDelegator(ClientConfig config) {
+    super(
+        emptyMap(),
+        Require.nonNull("HTTP client configuration", config),
+        getDefaultClientFactory());
+  }
+
+  public BiDiDelegator(
+      Map<String, CommandInfo> additionalCommands,
+      ClientConfig config,
+      HttpClient.Factory httpClientFactory) {
+    super(additionalCommands, config, httpClientFactory);
+  }
+
+  public BiDiDelegator(
+      Map<String, CommandInfo> additionalCommands, URL addressOfRemoteServer, ClientConfig config) {
+    super(
+        additionalCommands,
+        config.baseUrl(Require.nonNull("Server URL", addressOfRemoteServer)),
+        getDefaultClientFactory());
+  }
+
+  public void setBiDiCommandExecutor(BiDiCommandExecutor biDiCommandExecutor) {
+    this.biDiCommandExecutor = biDiCommandExecutor;
+  }
+
+  @Override
+  public Response execute(Command command) throws IOException {
+    if (biDiCommandExecutor != null && biDiCommands.contains(command.getName())) {
+      return biDiCommandExecutor.execute(command);
```

Comment:
If you sniff the response to see if it's an `unknown command` here, you could build the list of supported commands dynamically. This would allow the Bidi implementation to not be fully complete, and we'd fall back to Classic commands where possible.